### PR TITLE
fix(react-grid): correlate sorting order with grouping order

### LINKED ISSUE #398: TableHeaderRow does not cancel applied sorting within a scope
TableHeaderRow does not cancel applied sorting within a scope when a column in the same scope is clicked.

### PR DESCRIPTION
**Pull request by kvet**
fixes #398 

BREAKING CHANGES:

The `scope` parameter of the `setColumnSorting` action has been removed
 
The `GroupingState` plugin now has an optional dependency on the `SortingState` plugin. So, `GroupingState` should be placed after `SortingState`.

Before:

```jsx
<GroupingState /* ... */ />
<SortingState /* ... */ />
```

After:

```jsx
<SortingState /* ... */ />
<GroupingState /* ... */ />
```

### CODE REVIEW

File: packages/dx-grid-core/src/plugins/grouping-state/reducers.js [hunk 1]
```diff
@@ -47,5 +50,12 @@ export const toggleExpandedGroups = (prevExpandedGroups, { groupKey }) => {
     expandedGroups.push(groupKey);
   }
 
-  return expandedGroups;
+  return {
+    expandedGroups,
+  };
 };
+
+export const draftGroupingChange = (state, { columnName, groupIndex }) =>
```
Comment (SergeyAlexeev) [inline packages/dx-grid-core/src/plugins/grouping-state/reducers.js]: `state` isn't used

File: packages/dx-grid-core/src/plugins/sorting-state/reducers.js [hunk 2]
```diff
@@ -1,39 +1,37 @@
...
-  if (scope) {
-    nextSorting = [
-      ...nextSorting,
-      ...sorting.filter(sort =>
-        scope.indexOf(sort.columnName) === -1 &&
-        sort.columnName !== columnName),
-    ];
+  let nextSorting = [];
+  if (keepOther === true) {
+    nextSorting = sorting.slice();
```
Comment (SergeyAlexeev) [inline packages/dx-grid-core/src/plugins/sorting-state/reducers.js]: Check case when seamless immutable is used. Probably, `sorting.slice()` should be replaced to `Array.from`.

File: packages/dx-react-grid/src/plugins/grouping-state.jsx [hunk 3]
```diff
@@ -4,106 +4,135 @@ import { Getter, Action, PluginContainer } from '@devexpress/dx-react-core';
...
-      this.setState({
-        groupingChange: draftGroupingChange(this.state.groupingChange, groupingChange),
-      });
+    this.groupByColumn = this.groupByColumn.bind(this);
+    this.toggleGroupExpanded = this.applyReducer.bind(this, toggleExpandedGroups);
+    this.draftGroupingChange = this.applyReducer.bind(this, draftGroupingChange);
+    this.cancelGroupingChange = this.applyReducer.bind(this, cancelGroupingChange);
+    this.setColumnSorting = this.setColumnSorting.bind(this);
+  }
+  getState() {
```
Comment (dxbykov) [inline packages/dx-react-grid/src/plugins/grouping-state.jsx]: It's strange that 'get' function creates a new object on each call.

File: packages/dx-react-grid/src/plugins/grouping-state.jsx [hunk 4]
```diff
@@ -4,106 +4,135 @@ import { Getter, Action, PluginContainer } from '@devexpress/dx-react-core';
...
+      ...sorting[columnSortingIndex],
+    });
+  }
+  applyReducer(reduce, payload) {
+    const state = this.getState();
+    const nextState = reduce(state, payload);
+    this.setState(nextState);
+
+    const { onGroupingChange } = this.props;
+    if (onGroupingChange && nextState.grouping !== state.grouping) {
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid/src/plugins/grouping-state.jsx]: `nextState.grouping`, `nextState.expandedGroups` can be replaced to `const { grouping, expandedGroups } = nextState; `

File: packages/dx-react-grid/docs/reference/table-header-row.md [hunk 5]
```diff
@@ -64,7 +64,7 @@ tableHeaderRows | Getter | Array&lt;[TableRow](table-view.md#table-row)&gt; | He
 sorting | Getter | Array&lt;[Sorting](sorting-state.md#sorting)&gt; | Column sorting.
 columns | Getter | Array&lt;[Column](#column)&gt; | Table columns.
 grouping | Getter | Array&lt;[Grouping](grouping-state.md#grouping)&gt; | Columns used for grouping.
-setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean, cancel: boolean }) => void | Changes column sorting.
+setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean | Array&lt;String&gt;, cancel: boolean }) => void | Changes a column's sort direction. Keeps existing sorting if `keepOther` is set to `true`. The `keepOther` can handle contains the names of columns will be keeped when sorting is applied. Cancels sorting by the current column if `cancel` is set to `true`.
```
Comment (dxbykov) [inline packages/dx-react-grid/docs/reference/table-header-row.md]: keeped => kept

File: packages/dx-grid-core/src/plugins/sorting-state/reducers.test.js [hunk 6]
```diff
@@ -1,98 +1,156 @@
...
+          sorting: [],
+        });
+    });
 
-      const nextSorting = setColumnSorting(sorting, payload);
-      expect(nextSorting).toEqual([
-        { columnName: 'test', direction: 'asc' },
-        { columnName: 'test3', direction: 'asc' },
-      ]);
+    it('should set corretct sorting if sortIndex is specified', () => {
```
Comment (gsobolev) [inline packages/dx-grid-core/src/plugins/sorting-state/reducers.test.js]: Typo: 'corretct'

File: packages/dx-react-grid/docs/reference/grouping-panel.md [hunk 7]
```diff
@@ -89,7 +89,7 @@ columns | Getter | Array&lt;[Column](#column)&gt; | Grid columns.
 draftGrouping | Getter | Array&lt;[DraftGrouping](grouping-state.md#draft-grouping)&gt; | Grouping options used for the preview.
 sorting | Getter | Array&lt;[Sorting](sorting-state.md#sorting)&gt; | The current sorting state.
 groupByColumn | Action | ({ columnName: string }) => void | Toggles a column's grouping state.
-setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean, cancel: boolean }) => void | Updates column sorting.
+setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean | Array&lt;String&gt;, cancel: boolean }) => void | Changes a column's sort direction. Keeps existing sorting if `keepOther` is set to `true`. The `keepOther` can handle contains the names of columns will be kept when sorting is applied. Cancels sorting by the current column if `cancel` is set to `true`.
```
Comment (gsobolev) [inline packages/dx-react-grid/docs/reference/grouping-panel.md]: Please check this sentence
> The `keepOther` **can handle contains** the names...

File: packages/dx-react-grid/docs/reference/grouping-state.md [hunk 8]
```diff
@@ -55,6 +55,8 @@ A string value that consists of values by which rows are grouped, separated by t
 Name | Plugin | Type | Description
 -----|--------|------|------------
 columns | Getter | Array&lt;[Column](grid.md#column)&gt; | The grid columns.
+sorting? | Getter | Array&lt;[Sorting](sorting-state.md#sorting)&gt; | Applied column sorting.
+setColumnSorting? | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean | Array&lt;String&gt;, cancel: boolean }) => void | Changes a column's sort direction. Keeps existing sorting if `keepOther` is set to `true`. The `keepOther` can handle contains the names of columns will be kept when sorting is applied. Cancels sorting by the current column if `cancel` is set to `true`.
```
Comment (gsobolev) [inline packages/dx-react-grid/docs/reference/grouping-state.md]: > can handle contains

File: packages/dx-react-grid/docs/reference/sorting-state.md [hunk 9]
```diff
@@ -40,4 +40,4 @@ none
 Name | Plugin | Type | Description
 -----|--------|------|------------
 sorting | Getter | Array&lt;[Sorting](#sorting)&gt; | Applied column sorting.
-setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean, cancel: boolean, scope: Array&lt;String&gt; }) => void | Changes a column's sort direction. Keeps existing sorting if `keepOther` is set to `true`. Cancels sorting by the current column if `cancel` is set to `true`. The `scope` array contains the names of columns taken into account when sorting.
+setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean | Array&lt;String&gt;, cancel: boolean }) => void | Changes a column's sort direction. Keeps existing sorting if `keepOther` is set to `true`. The `keepOther` can handle contains the names of columns will be kept when sorting is applied. Cancels sorting by the current column if `cancel` is set to `true`.
```
Comment (gsobolev) [inline packages/dx-react-grid/docs/reference/sorting-state.md]: > can handle contains

File: packages/dx-react-grid/docs/reference/table-header-row.md [hunk 10]
```diff
@@ -64,7 +64,7 @@ tableHeaderRows | Getter | Array&lt;[TableRow](table-view.md#table-row)&gt; | He
 sorting | Getter | Array&lt;[Sorting](sorting-state.md#sorting)&gt; | Column sorting.
 columns | Getter | Array&lt;[Column](#column)&gt; | Table columns.
 grouping | Getter | Array&lt;[Grouping](grouping-state.md#grouping)&gt; | Columns used for grouping.
-setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean, cancel: boolean }) => void | Changes column sorting.
+setColumnSorting | Action | ({ columnName: string, direction: 'asc' &#124; 'desc', keepOther: boolean | Array&lt;String&gt;, cancel: boolean }) => void | Changes a column's sort direction. Keeps existing sorting if `keepOther` is set to `true`. The `keepOther` can handle contains the names of columns will be kept when sorting is applied. Cancels sorting by the current column if `cancel` is set to `true`.
```
Comment (gsobolev) [inline packages/dx-react-grid/docs/reference/table-header-row.md]: > can handle contains

File: packages/dx-react-grid/src/plugins/local-paging.test.jsx [hunk 11]
```diff
@@ -81,8 +81,8 @@ describe('LocalPaging', () => {
       </PluginHost>,
     );
 
-    expect(defaultDeps.action.setCurrentPage.mock.calls)
-      .toEqual([[2]]);
+    expect(defaultDeps.action.setCurrentPage.mock.calls[0][0])
+      .toEqual(2);
```
Comment (gsobolev) [inline packages/dx-react-grid/src/plugins/local-paging.test.jsx]: Why not use the `.toHaveBeenCalledWith()` matcher?

Comment (kvet) [reply packages/dx-react-grid/src/plugins/local-paging.test.jsx]: Action is called with additional arguments (getters and actions)

**Pull request merged**